perception/autoware_lidar_centerpoint/lib/preprocess/preprocess_kernel.cu

```diff
@@ -185,12 +185,14 @@ __global__ void generateBaseFeatures_kernel(
   unsigned int * mask, float * voxels, int grid_y_size, int grid_x_size, int max_voxel_size,
   unsigned int * pillar_num, float * voxel_features, float * voxel_num, int * voxel_idxs)
 {
-  unsigned int voxel_idx = blockIdx.x * blockDim.x + threadIdx.x;
-  unsigned int voxel_idy = blockIdx.y * blockDim.y + threadIdx.y;
-
-  if (voxel_idx >= grid_x_size || voxel_idy >= grid_y_size) return;
-
-  unsigned int voxel_index = voxel_idy * grid_x_size + voxel_idx;
+  // exchange x and y to process in a row-major order
+  // flip x axis direction to process front to back
+  unsigned int voxel_idx_inverted = blockIdx.y * blockDim.y + threadIdx.y;
+  unsigned int voxel_idy = blockIdx.x * blockDim.x + threadIdx.x;
+  if (voxel_idx_inverted >= grid_x_size || voxel_idy >= grid_y_size) return;
+  unsigned int voxel_idx = grid_x_size - 1 - voxel_idx_inverted;
+
+  unsigned int voxel_index = voxel_idx_inverted * grid_y_size + voxel_idy;
```

Comment:
If I understand this correctly, the addresses are in decreasing order. For optimal access, addresses should be in increasing order (bursts of memory accesses).

Reply:
The idea is that the voxel_index is inverted at `generateVoxels_random_kernel`
```c++
int voxel_idx = floorf((point.x - min_x_range) / pillar_x_size);
int voxel_idy = floorf((point.y - min_y_range) / pillar_y_size);
unsigned int voxel_index = (grid_x_size - 1 - voxel_idx) * grid_y_size + voxel_idy;
```
the voxel_index becomes 0 at (`voxel_idx=grid_x_size-1`, `voxel_idy=0`), front-right of the ego.

so that `generateBaseFeatures_kernel` can access memory in increasing order.
```c++
unsigned int voxel_idx_inverted = blockIdx.y * blockDim.y + threadIdx.y;
unsigned int voxel_idy = blockIdx.x * blockDim.x + threadIdx.x;

unsigned int voxel_index = voxel_idx_inverted * grid_y_size + voxel_idy;
```
It processes from the front-right to front-left, front row to rear row.

If my understand is incorrect, can you show an example?

Reply:
No, you are right, I read wrong how was `voxel_idx_inverted` was being used now :bow:

I had the impression that although the data was being read by row (memory friendly), it was being from the highest value in each row to the lowest, which actually due to how warps gather its threads' reads it is not as bad as in cpu